homeassistant/components/reolink/host.py

```diff
@@ -162,4 +179,140 @@ async def disconnect(self):
+            )
+
+        return True
+
+    async def register_webhook(self) -> bool:
+        """Register the webhook for motion events."""
+        self.webhook_id = f"reolink_{self.unique_id.replace(':', '')}"
+        event_id = self.webhook_id
+
+        try:
+            webhook.async_register(
+                self._hass, DOMAIN, event_id, self.webhook_id, self.handle_webhook
+            )
+        except ValueError:
+            _LOGGER.debug(
+                "Error registering webhook %s. Trying to unregister it first and re-register again",
+                self.webhook_id,
+            )
+            webhook.async_unregister(self._hass, self.webhook_id)
+            try:
+                webhook.async_register(
+                    self._hass, DOMAIN, event_id, self.webhook_id, self.handle_webhook
+                )
+            except ValueError:
+                _LOGGER.error(
+                    "Error registering a webhook %s for %s",
+                    self.webhook_id,
+                    self.api.nvr_name,
+                )
+                self.webhook_id = None
+                return False
+
+        try:
+            base_url = get_url(self._hass, prefer_external=False)
+        except NoURLAvailableError:
+            try:
+                base_url = get_url(self._hass, prefer_external=True)
+            except NoURLAvailableError:
+                _LOGGER.error(
+                    "Error registering URL for webhook %s: HomeAssistant URL is not available",
+                    self.webhook_id,
+                )
+                webhook.async_unregister(self._hass, self.webhook_id)
+                self.webhook_id = None
+                return False
+
+        webhook_path = webhook.async_generate_path(self.webhook_id)
+        self._webhook_url = f"{base_url}{webhook_path}"
+
+        _LOGGER.info("Registered webhook: %s", self.webhook_id)
+        return True
+
+    async def unregister_webhook(self):
+        """Unregister the webhook for motion events."""
+        if self.webhook_id:
+            _LOGGER.info("Unregistering webhook %s", self.webhook_id)
+            webhook.async_unregister(self._hass, self.webhook_id)
+        self.webhook_id = None
+
+    async def handle_webhook(
+        self, hass: HomeAssistant, webhook_id: str, request: Request
+    ):
+        """Handle incoming webhook from Reolink for inbound messages and calls."""
+
+        _LOGGER.debug("Webhook '%s' called", webhook_id)
+
+        if not request.body_exists:
+            _LOGGER.debug("Webhook '%s' triggered without payload", webhook_id)
+
+        data = await request.text()
+        if not data:
+            _LOGGER.debug(
+                "Webhook '%s' triggered with unknown payload: %s", webhook_id, data
+            )
+            return
+
+        channel = await self._api.ONVIF_event_callback(data)
+
+        if channel is None:
+            hass.bus.async_fire(f"{webhook_id}_all", {})
```

Comment:
Don't use the event bus for internal communication. Use our dispatch helper to signal the entities instead. Only user facing events should be passed to the event bus from integrations.

Here's how the netatmo integrations does it:

https://github.com/home-assistant/core/blob/941f82b60c405dd1173e892d395ee47a4cc76318/homeassistant/components/netatmo/webhook.py#L76-L99

You can see that only specific webhook events are marked as user facing events and fired on the event bus, while all webhook events are dispatched internally to listeners inside the integration.

Reply:
@MartinHjelmare thank you very much, I keep on learning, this is much better.